server/athenian/api/internal/miners/github/deployment.py

```diff
@@ -2025,3 +2028,142 @@ async def load_jira_issues_for_deployments(deployments: pd.DataFrame,
+        for name, name_groups in groupby(
+            sorted(outlier_deploys, key=attrgetter("deployment_name")),
+            key=attrgetter("deployment_name"),
+        )
+    ]
+
+    tables = (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
+    for Table in tables:
+        where_clause = sa.and_(
+            Table.acc_id == account,
+            sa.or_(
+                *[
+                    sa.and_(Table.deployment_name == g[0], Table.repository_full_name.in_(g[1]))
+                    for g in grouped_deploys
+                ],
+            ),
+        )
+        stmts.append(sa.delete(Table).where(where_clause))
+
+    # clear GitHubDeploymentFacts by removing prs and commits from facts data
+    FactsT = GitHubDeploymentFacts
+    depl_facts_stmt = sa.select(GitHubDeploymentFacts).where(
+        sa.and_(
+            FactsT.acc_id == account,
+            FactsT.deployment_name.in_(d.deployment_name for d in outlier_deploys),
+        ),
+    )
+    depl_facts_rows = await pdb.fetch_all(depl_facts_stmt)
+
+    for row in depl_facts_rows:
+        facts = DeploymentFacts(row[FactsT.data.name], name=row[FactsT.deployment_name.name])
+        new_facts = facts.with_nothing_deployed()
+        match_cols = (
+            FactsT.acc_id,
+            FactsT.deployment_name,
+            FactsT.release_matches,
+            FactsT.format_version,
+        )
+        update_stmt = sa.update(FactsT).where(
+            sa.and_(*(col == row[col.name] for col in match_cols)),
+        ).values({FactsT.data: new_facts.data})
+        stmts.append(update_stmt)
+
+    await gather(*[pdb.execute(stmt) for stmt in stmts], op="hide_first_deployments")
+
+
+@dataclass(frozen=True, slots=True)
+class _OutlierDeployment:
+    repository_full_name: str
+    deployment_name: str
+
+
+async def _search_outlier_first_deployments(
+    deployment_facts: pd.DataFrame,
+    meta_ids: Sequence[int],
+    mdb: Database,
+    threshold: float = 100.0,
+) -> Sequence[_OutlierDeployment]:
+    """Search the outlier first deployments."""
+    exploded_facts = deployment_facts.explode("repositories").rename(
+        columns={"repositories": "repository"},
+    )
+
+    # retrieve repo => creation time mapping
+    mentioned_repos = exploded_facts.repository.unique()
+    repos_stmt = sa.select(Repository.full_name, Repository.created_at).where(
+        sa.and_(Repository.acc_id.in_(meta_ids), Repository.full_name.in_(mentioned_repos)),
+    )
+    repo_creation_times = {
+        r[0]: ensure_db_datetime_tz(r[1], mdb)
+        for r in await mdb.fetch_all(repos_stmt)
+    }
+
+    grouped_facts = exploded_facts.groupby(["environment", "repository"])
+
+    result = set()
+
+    log = logging.getLogger(f"{__name__}.hide_outlier_first_deployments")
+    for group_name, group in grouped_facts:
+        env, repo = group_name
+        sorted_group = group.sort_values(["started_at"], axis="index")
```

Comment:
Algorithm:

Take the earliest deployment by `argmin()`. You don't have to sort now.

To calculate the median, this code will execute 10-100x faster (Pandas sucks):

```python
successful_mask = group[DeploymentNotification.conclusion.name].values == DeploymentNotification.CONCLUSION_SUCCESS
value = np.median(group[DeploymentNotification.started_at.name].values[successful_mask])
```

Reply:
I think I have to sort anyway to compute the median of the diffs of started_at
I've rewritten it to work less on the pandas level and more with numpy

Reply:
OK!